dragonfly/session/handler_player_action.go

```diff
@@ -55,11 +56,26 @@ func handlePlayerAction(action int32, face int32, pos protocol.BlockPos, entityR
 		s.c.AbortBreaking()
 	case protocol.PlayerActionStopBreak:
 		s.c.FinishBreaking()
+	case protocol.PlayerActionContinueDestroyBlock:
+		if s.c.Breaking() {
+			s.c.AbortBreaking()
+			s.c.StartBreaking(world.BlockPos{int(pos[0]), int(pos[1]), int(pos[2])}, world.Face(face))
+		}
+		fallthrough
 	case protocol.PlayerActionCrackBreak:
 		s.swingingArm.Store(true)
 		defer s.swingingArm.Store(false)
 
 		s.c.ContinueBreaking(world.Face(face))
+	case protocol.PlayerActionPredictDestroyBlock:
+		s.swingingArm.Store(true)
+		defer s.swingingArm.Store(false)
+
+		s.c.StartBreaking(world.BlockPos{int(pos[0]), int(pos[1]), int(pos[2])}, world.Face(face))
+		s.c.FinishBreaking()
+		if _, ok := s.c.GameMode().(gamemode.Survival); ok {
+			s.c.StartBreaking(world.BlockPos{int(pos[0]), int(pos[1]), int(pos[2])}.Side(world.Face(face).Opposite()), world.Face(face))
```

Comment:
I'm a bit concerned over this part. We can't just take the opposite face, because if a player is first destroying the top of a block at an angle, when that block breaks it'll target the one on the side of it, not under it.

Reply:
Changed